app/metal-controller-manager/controllers/environment_controller.go

```diff
@@ -93,43 +94,42 @@ func (r *EnvironmentReconciler) reconcile(req ctrl.Request) (ctrl.Result, error)
 	} {
 		assetTask := assetTask
 
-		file := filepath.Join(envs, assetTask.BaseName)
+		file = filepath.Join(envs, assetTask.BaseName)
 
-		if _, err := os.Stat(file); os.IsNotExist(err) {
-			wg.Add(1)
-
-			go func() {
-				defer wg.Done()
-
-				l.Info("saving asset", "url", assetTask.Asset.URL)
+		setReady := func(ready bool) {
+			status := "False"
+			if ready {
+				status = "True"
+			}
 
-				if err := save(assetTask.Asset, file); err != nil {
-					condition := metalv1alpha1.AssetCondition{
-						Asset:  assetTask.Asset,
-						Status: "False",
-						Type:   "Ready",
-					}
+			condition := metalv1alpha1.AssetCondition{
+				Asset:  assetTask.Asset,
+				Status: status,
+				Type:   "Ready",
+			}
 
-					mu.Lock()
-					conditions = append(conditions, condition)
-					mu.Unlock()
+			mu.Lock()
+			conditions = append(conditions, condition)
+			mu.Unlock()
+		}
 
+		saveAsset := func() {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				if err := save(ctx, assetTask.Asset, file); err != nil {
```

Comment:
this is going to be a problem - file is outer variable which gets changes in the for loop and used in goroutine, so it will change the value while goroutine is running

Reply:
ah, right. I've missed that wg is being waited after the loop.

Reply:
fixed